README.rst
QR Code generator for Inkscape
==============================

While Inkscape provides a QR Code generator, this extension provides slightly
more options (i.e. generation of Micro QR Codes and an optimal error correction
level) and should be standard-conform.

This Inkscape extension includes `Segno <https://github.com/heuer/segno/>`_, a
BSD-licensed, pure Python library which implements the ``ISO/IEC 18004:2015(E)``
standard for QR Codes.


Installation
------------

Use ``pip`` to install inkscape-qrcode from PyPI::

    $ pip install --update --target=<target_dir> inkscape-qrcode

``<target_dir>`` should refer to an absolute path, see below for typical
installation directories.

Alternatively copy the ``inkscape_qrcode.inx`` and the ``inkscape_qrcode``
directory into the target directoy.

The QR Code generator should appear in the menu
``Extensions > Render > Barcode > QR Code (Segno)``



Inkscape extension path
-----------------------

OS X
^^^^
* global: /Applications/Inkscape.app/Contents/Resources/extensions
* local: $HOME/.config/inkscape/extensions


Linux
^^^^^
* global: /usr/share/inkscape/extensions
* local: $HOME/.config/inkscape/extensions


Windows
^^^^^^^
* global: C:\\Program Files\\Inkscape\\share\\extensions
* local: C:\\Documents and Settings\\User\\Application Data\\Inkscape\\extensions


CHANGES.rst
Changes
=======


0.1.7 - 2018-03-23
------------------
* Attempt to make the package installable via pip >= 7 again.


0.1.6 - 2018-03-21
------------------
* Updated docs and provide an additional zip file


0.1.5 - 2018-03-13
------------------
* Updated Segno to version 0.2.7
* Fixed #1 "Support for Structured Append"
* Fixed #2 (avoid exceptions if a Micro QR Code is chosen)


0.1.4 - 2017-01-31
------------------
* Updated Segno version to 0.2.4 to avoid problems with Micro QR Codes


0.1.3 - 2016-09-15
------------------
* Fixed inkscape_qrcode.inx (pointed to a wrong path)
* Fixed Python packaging


0.1.2 - 2016-09-15
------------------
* Updated underlying Segno version to 0.2.0 since
  the previous version may create invalid QR Codes if "error boost" is
  enabled


0.1.1 - 2016-09-07
------------------
* Initial release


update_segno.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Inkscape QR Code
# Copyright (C) 2016 - 2018 Lars Heuer
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public
# License as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
"""\
Helper script to update Segno to the latest version.
"""
from __future__ import absolute_import, unicode_literals, print_function
import json
import os
import re
import io
import tarfile
import shutil
try:
    unicode
    from io import open
except NameError:
    pass
try:
    from urllib.request import urlopen
except ImportError:
    from urllib2 import urlopen


_SEGNO_URL = 'https://pypi.python.org/pypi/segno/json'
_LINK_PATTERN = re.compile(r'<a href="([^"]+)"[^>]*>segno\-[0-9\.]+tar\.gz')
_VERSION_PATTERN = re.compile(r'''__version__\s+=\s+['"]([0-9\.]+)['"]''')
_VERSION_PATTERN_FILENAME = re.compile('segno-([0-9\.]+)\.tar\.gz')

_WANTED_FILES = ('consts.py', 'encoder.py', 'utils.py')


def update_segno():
    """\
    Downloads the latest Segno version and updates the "_segno" package.
    """
    segno_subdir = os.path.join(os.path.dirname(__file__), 'inkscape_qrcode', '_segno')
    with open(os.path.join(segno_subdir, '__init__.py')) as f:
        m = _VERSION_PATTERN.search((f.read()))
    segno_internal_version = m.group(1)
    resp = urlopen(_SEGNO_URL)
    if resp.getcode() != 200:
        raise Exception('Error retrieving the PyPi page: %s' % resp.getcode())
    j = json.loads(resp.read())
    download_url = None
    for urldct in j['urls']:
        if urldct['python_version'] == 'source':
            download_url = urldct['url']
            break
    current_version = _VERSION_PATTERN_FILENAME.search(download_url).group(1)
    if segno_internal_version == current_version:
        print('No update found. Current Segno version: "%s"' % segno_internal_version)
        return
    resp = urlopen(download_url)
    if resp.getcode() != 200:
        raise Exception('Download of package failed: %s' % resp.getcode())
    buff = io.BytesIO(resp.read())
    buff.seek(0)
    needed_files = list(_WANTED_FILES)
    with tarfile.open(fileobj=buff, mode='r:gz') as f:
        for member in f.getmembers():
            member_name = os.path.basename(member.name)
            if member_name in needed_files:
                mf = f.extractfile(member)
                needed_files.remove(member_name)
                with open(os.path.join(segno_subdir, member_name), 'wb') as fdst:
                    shutil.copyfileobj(mf, fdst)
                if not needed_files:
                    break
    if needed_files:
        print('Not all files found. Missing: %r' % needed_files)
        print("Don't forget to revert changes!")
        return
    segno_init_file = os.path.join(segno_subdir, '__init__.py')
    with open(segno_init_file, 'r') as f:
        s = f.read()
    s = s.replace(segno_internal_version, current_version)
    with open(segno_init_file, 'w') as f:
        f.write(s)
    inx_fn = os.path.join(os.path.dirname(__file__), 'inkscape_qrcode.inx')
    with open(inx_fn, 'r', encoding='utf-8') as f:
        s = f.read()
    s = s.replace('Segno %s' % segno_internal_version, 'Segno %s' % current_version)
    with open(inx_fn, 'w', encoding='utf-8') as f:
        f.write(s)
    print('Updated Segno from %s to %s' % (segno_internal_version, current_version))


if __name__ == '__main__':
    update_segno()


inkscape_qrcode.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Inkscape QR Code
# Copyright (C) 2016 Lars Heuer
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public
# License as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
#  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
"""\
Inkscape extension which uses Segno to generate QR Codes.
"""
from __future__ import absolute_import, unicode_literals
import io
import inkex
try:
    from simpletransform import computePointInNode
except ImportError:
    def computePointInNode(pt, node):
        return pt
try:
    from ._segno import encoder, utils
except (ValueError, ImportError, SystemError):
    from _segno import encoder, utils

__version__ = '0.1.0'


class InkscapeQRCode(inkex.Effect):
    """\
    This class mediates between Inkscape and Segno.
    """
    def __init__(self):
        inkex.Effect.__init__(self)
        self.OptionParser.add_option('--data', action='store',
                                     dest='data', type='string')
        self.OptionParser.add_option('--version', action='store',
                                     dest='version', type='string')
        self.OptionParser.add_option('--scale', action='store',
                                     dest='scale', type='float', default=1.0)
        self.OptionParser.add_option('--error', action='store',
                                     dest='error', type='string', default='m')
        # Actually these are booleans but we keep them as str
        self.OptionParser.add_option('--background', action='store',
                                     dest='background', type='string',
                                     default='false')
        self.OptionParser.add_option('--allow_micro', action='store',
                                     dest='micro', type='string',
                                     default='false')
        self.OptionParser.add_option('--boost_error', action='store',
                                     dest='boost_error', type='string',
                                     default='false')
        
    def effect(self):
        opts = self.options
        version, error = opts.version, opts.error
        if version.upper() == 'M1' and error:
            error = None
        if version == '-':
            version = None
        micro = None if opts.micro == 'true' else False
        boost_error = opts.boost_error == 'true'
        want_background = opts.background == 'true'
        
        qr = encoder.encode(opts.data, version=version, error=error,
                            micro=micro, boost_error=boost_error)

        border = utils.get_default_border_size(qr.version)

        # Create path data
        x, y = border, border + .5  # .5 == stroke-width / 2
        line_iter = utils.matrix_to_lines(qr.matrix, x, y)
        # 1st coord is absolute
        (x1, y1), (x2, y2) = next(line_iter)
        coord = ['M{0} {1}h{2}'.format(x1, y1, x2 - x1)]
        append_coord = coord.append
        x, y = x2, y2
        for (x1, y1), (x2, y2) in line_iter:
            append_coord('m{0} {1}h{2}'.format(x1 - x, int(y1 - y), x2 - x1))
            x, y = x2, y2
        path_data = ''.join(coord)

        centre = tuple(computePointInNode(list(self.view_center), self.current_layer))

        grp_transform = 'translate' + str(centre)
        if opts.scale != 1:
            grp_transform += ' scale(%f)' % opts.scale
        grp = inkex.etree.SubElement(self.current_layer, inkex.addNS('g', 'svg'),
                                     transform=grp_transform)
        if want_background:
            width, height = utils.get_symbol_size(qr.version, border=border)
            inkex.etree.SubElement(grp, inkex.addNS('rect', 'svg'),
                                   width=str(width), height=str(height),
                                   fill='#FFF')
        inkex.etree.SubElement(grp, inkex.addNS('path', 'svg'),
                               d=path_data, stroke='#000')
        

if __name__ == '__main__':
    e = InkscapeQRCode()
    e.affect()


inkscape_qrcode/qrcode.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Inkscape QR Code
# Copyright (C) 2016 - 2018 Lars Heuer
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public
# License as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
"""\
Inkscape extension which uses Segno to generate QR Codes.
"""
from __future__ import absolute_import, unicode_literals
import inkex
try:
    from ._segno import encoder, utils
except (ImportError, ValueError):
    from _segno import encoder, utils
try:
    from simpletransform import computePointInNode
except ImportError:
    def computePointInNode(pt, node):
        return pt

__version__ = '0.1.8.dev0'


class InkscapeQRCode(inkex.Effect):
    """\
    This class mediates between Inkscape and Segno.
    """
    def __init__(self):
        inkex.Effect.__init__(self)
        self.OptionParser.add_option('--data', action='store',
                                     dest='data', type='string', default="")
        self.OptionParser.add_option('--version', action='store',
                                     dest='version', type='string')
        self.OptionParser.add_option('--scale', action='store',
                                     dest='scale', type='float', default=1.0)
        self.OptionParser.add_option('--error', action='store',
                                     dest='error', type='string', default='L')
        self.OptionParser.add_option('--symbol_count', action='store',
                                     dest='symbol_count',  type='int', default=1)
        # Actually these are booleans but we keep them as str
        self.OptionParser.add_option('--background', action='store',
                                     dest='background', type='string', default='false')
        self.OptionParser.add_option('--allow_micro', action='store',
                                     dest='micro', type='string', default='false')
        self.OptionParser.add_option('--boost_error', action='store',
                                     dest='boost_error', type='string', default='false')

    def effect(self):
        opts = self.options
        version, error = opts.version, opts.error
        if version.upper() == 'M1' and error:
            error = None
        if version == '-':
            version = None
        micro = None if opts.micro == 'true' else False
        # Avoid problems / exceptions with choosing a Micro QR code version
        # If micro is False and the user chooses a MQR, an exception will be raised
        if version in ('M1', 'M2', 'M3', 'M4'):
            micro = True
        boost_error = opts.boost_error == 'true'
        want_background = opts.background == 'true'
        if not micro and opts.symbol_count > 1:
            qrs = encoder.encode_sequence(opts.data, version=version, error=error,
                                          encoding=encoder, boost_error=boost_error,
                                          symbol_count=opts.symbol_count)
        else:
            qrs = [encoder.encode(opts.data, version=version, error=error,
                                  micro=micro, boost_error=boost_error)]
        centre = tuple(computePointInNode(list(self.view_center), self.current_layer))
        grp_transform = 'translate' + str(centre)
        if opts.scale != 1:
            grp_transform += ' scale(%f)' % opts.scale
        grp = inkex.etree.SubElement(self.current_layer, inkex.addNS('g', 'svg'),
                                     transform=grp_transform)
        border = utils.get_default_border_size(qrs[0].version)
        width, height = utils.get_symbol_size(qrs[0].version, border=border)
        multiple_qr = len(qrs) > 1
        offset = 0
        for qr in qrs:
            g = grp if not multiple_qr else inkex.etree.SubElement(grp, inkex.addNS('g', 'svg'))
            if want_background:
                inkex.etree.SubElement(g, inkex.addNS('rect', 'svg'),
                                       width=str(width), height=str(height),
                                       x=str(offset), fill='#FFF')
            path_data = _create_path(qr, border, offset=offset)
            inkex.etree.SubElement(g, inkex.addNS('path', 'svg'), d=path_data,
                                   stroke='#000')
            offset += width + border


def _create_path(qr, border, offset):
    """\
    Returns a path of dark modules.

    :param qr: QR code
    :param int border: The border size
    """
    # Create path data
    x, y = border + offset, border + .5  # .5 == stroke-width / 2
    line_iter = utils.matrix_to_lines(qr.matrix, x, y)
    # 1st coord is absolute
    (x1, y1), (x2, y2) = next(line_iter)
    coord = ['M{0} {1}h{2}'.format(x1, y1, x2 - x1)]
    append_coord = coord.append
    x, y = x2, y2
    for (x1, y1), (x2, y2) in line_iter:
        append_coord('m{0} {1}h{2}'.format(x1 - x, int(y1 - y), x2 - x1))
        x, y = x2, y2
    return ''.join(coord)


if __name__ == '__main__':
    e = InkscapeQRCode()
    e.affect()
